samza-kafka/src/main/java/org/apache/samza/system/kafka/SamzaKafkaSystemAdmin.java

```diff
@@ -0,0 +1,680 @@
+      allNewestOffsets.putAll(offsetsForTopic.getNewestOffsets());
+      allUpcomingOffsets.putAll(offsetsForTopic.getUpcomingOffsets());
+    });
+
+    scala.collection.immutable.Map<String, SystemStreamMetadata> result =
+        KafkaSystemAdminUtilsScala.assembleMetadata(ScalaJavaUtil.toScalaMap(allOldestOffsets),
+            ScalaJavaUtil.toScalaMap(allNewestOffsets), ScalaJavaUtil.toScalaMap(allUpcomingOffsets));
+
+    LOG.info("assembled SystemStreamMetadata is: {}", result);
+    return JavaConverters.mapAsJavaMapConverter(result).asJava();
+  }
+
+  private String fetchNewestOffset(SystemStreamPartition ssp) {
+    LOG.debug("Fetching newest offset for {}", ssp);
+    String newestOffset;
+
+    // create a kafka consumer for fetching meta data
+    TopicPartition topicPartition = new TopicPartition(ssp.getStream(), ssp.getPartition().getPartitionId());
+    List<TopicPartition> topicPartitions = Collections.singletonList(topicPartition);
+
+    // the offsets returned from the consumer is the Long type
+    Long upcomingOffset = (Long) metadataConsumer.endOffsets(topicPartitions).get(topicPartition);
+
+    // Kafka's "latest" offset is always last message in stream's offset + 1,
+    // so get newest message in stream by subtracting one. this is safe
+    // even for key-deduplicated streams, since the last message will
+    // never be deduplicated.
+    if (upcomingOffset <= 0) {
+      LOG.debug("Stripping newest offsets for {} because the topic appears empty.", topicPartition);
+      newestOffset = null;
+    } else {
+      newestOffset = String.valueOf(upcomingOffset - 1);
+    }
+
+    LOG.debug("Newest offset for ssp {} is: {}", ssp, newestOffset);
+    return newestOffset;
+  }
+
+  @Override
+  public Integer offsetComparator(String offset1, String offset2) {
+    if (offset1 == null || offset2 == null) {
+      return -1;
+    }
+
+    return Long.valueOf(offset1).compareTo(Long.valueOf(offset2));
+  }
+
+  @Override
+  public boolean createStream(StreamSpec streamSpec) {
+    LOG.info("Creating Kafka topic: {} on system: {}", streamSpec.getPhysicalName(), streamSpec.getSystemName());
+
+    return kafkaAdminUtils.createStream(toKafkaSpec(streamSpec), connectZk);
+  }
+
+  @Override
+  public boolean clearStream(StreamSpec streamSpec) {
+    LOG.info("Creating Kafka topic: {} on system: {}", streamSpec.getPhysicalName(), streamSpec.getSystemName());
+
+    kafkaAdminUtils.clearStream(streamSpec, connectZk);
+
+    Map<String, List<PartitionInfo>> topicsMetadata = getTopicMetadata(ImmutableSet.of(streamSpec.getPhysicalName()));
+    return topicsMetadata.get(streamSpec.getPhysicalName()).isEmpty();
+  }
+
+  /**
+   * Converts a StreamSpec into a KafakStreamSpec. Special handling for coordinator and changelog stream.
+   * @param spec a StreamSpec object
+   * @return KafkaStreamSpec object
+   */
+  KafkaStreamSpec toKafkaSpec(StreamSpec spec) {
+    KafkaStreamSpec kafkaSpec;
+    if (spec.isChangeLogStream()) {
```

Comment:
KafkaSystemAdmin should not concern itself with coordinator stream properties, changelog properties, intermediate stream properties. These streams are generally high level samza concepts and the underlying IO should not concern itself with constructing different properties bag for each stream. It's the responsibilty of the caller to pass in appropriate properties for each stream type before creating them.